File: lib/src/calendar.js
/**
 * Calendar, Schedule, Period, and Interval classes.
 * Uses Temporal throughout. No localStorage, no module-level globals.
 */

import { parsePlainDate, resolveScheduleTimes, daysBetween, noon, includesWeekend } from './datetime.js';

/**
 * Normalize the includeTags option.
 * Accepts either a flat array (same tags every weekday) or a per-day-of-week map.
 * Returns a map: { 1: [...], 2: [...], 3: [...], 4: [...], 5: [...] }
 *
 * @param {string[] | Record<number, string[]>} includeTags
 * @returns {Record<number, string[]>}
 */
const normalizeIncludeTags = (includeTags) => {
  if (Array.isArray(includeTags)) {
    return { 1: includeTags, 2: includeTags, 3: includeTags, 4: includeTags, 5: includeTags };
  }
  return includeTags || {};
};

class Calendar {
  #noonInstant(date) {
    return noon(date).toZonedDateTime(this.timezone).toInstant();
  }

  #nextSchoolDay(date) {
    let d = date.add({ days: 1 });
    while (!this.isSchoolDay(d)) {
      d = d.add({ days: 1 });
    }
    return d;
  }

  #previousSchoolDay(date) {
    let d = date.subtract({ days: 1 });
    while (!this.isSchoolDay(d)) {
      d = d.subtract({ days: 1 });
    }
    return d;
  }

  #schoolTimeBetween(start, end) {
    // Clamp start/end to calendar bounds.
    const calStart = this.startOfYear();
    const calEnd = this.endOfYear();
    const cursor = Temporal.Instant.compare(start, calStart) < 0 ? calStart : start;
    const finish = Temporal.Instant.compare(end, calEnd) > 0 ? calEnd : end;

    if (Temporal.Instant.compare(cursor, finish) >= 0) {
      return Temporal.Duration.from({ seconds: 0 });
    }

    let totalNs = 0n;
    let cursorDate = cursor.toZonedDateTimeISO(this.timezone).toPlainDate();

    while (Temporal.PlainDate.compare(cursorDate, this.lastDay) <= 0) {
      // Stop once the start of this day is past the finish time.
      const dayMidnightInstant = cursorDate.toPlainDateTime({ hour: 0 })
        .toZonedDateTime(this.timezone).toInstant();
      if (Temporal.Instant.compare(dayMidnightInstant, finish) >= 0) break;

      if (this.isSchoolDay(cursorDate)) {
        const sched = this.schedule(cursorDate);
        const dayStart = sched.startOfDay(cursorDate, this.timezone);
        const dayEnd = sched.endOfDay(cursorDate, this.timezone);

        const from = Temporal.Instant.compare(cursor, dayStart) > 0 ? cursor : dayStart;
        const to = Temporal.Instant.compare(finish, dayEnd) < 0 ? finish : dayEnd;

        if (Temporal.Instant.compare(from, to) < 0) {
          totalNs += to.epochNanoseconds - from.epochNanoseconds;
        }
      }

      cursorDate = cursorDate.add({ days: 1 });
    }

    const totalSeconds = Number(totalNs / 1_000_000_000n);
    const minutes = Math.floor(totalSeconds / 60);
    const hours = Math.floor(minutes / 60);

    return Temporal.Duration.from({ hours, minutes: minutes % 60, seconds: totalSeconds % 60 });
  }

  /**
   * @param {object} data - One year's calendar data object.
   * @param {{ role: string, includeTags: Record<number, string[]> }} options
   */
  constructor(data, options) {
    this.data = data;
    this.timezone = data.timezone;
    this.role = options.role;
    this.includeTags = normalizeIncludeTags(options.includeTags);

    this.firstDay = parsePlainDate(
      this.role === 'teacher' && data.firstDayTeachers ? data.firstDayTeachers : data.firstDay
    );
    this.lastDay = parsePlainDate(data.lastDay);
    this.schedules = data.schedules;
    this.holidays = data.holidays || [];
    this.teacherWorkDays = data.teacherWorkDays || [];
    this.breakNames = data.breakNames || {};
  }

  /**
   * @param {Temporal.Instant} instant
   * @returns {boolean}
   */
  isInCalendar(instant) {
    return (
      Temporal.Instant.compare(this.startOfYear(), instant) <= 0 &&
      Temporal.Instant.compare(instant, this.endOfYear()) <= 0
    );
  }

  startOfYear() {
    const sched = this.schedule(this.firstDay);
    return sched.firstPeriod().startInstant(this.firstDay, this.timezone);
  }

  endOfYear() {
    const sched = this.schedule(this.lastDay);
    return sched.lastPeriod().endInstant(this.lastDay, this.timezone);
  }

  /**
   * @param {Temporal.PlainDate} date
   * @returns {Schedule}
   */
  schedule(date) {
    const d = date.toString();
    let periods;
    if (d in this.schedules && d >= this.firstDay.toString()) {
      periods = this.schedules[d];
    } else if (date.dayOfWeek === 1) {
      // Monday: late start
      periods = this.schedules['default'].LATE_START;
    } else {
      periods = this.schedules['default'].NORMAL;
    }
    return new Schedule(this, resolveScheduleTimes(periods), date);
  }

  /**
   * @param {Temporal.PlainDate} date
   * @returns {boolean}
   */
  isSchoolDay(date) {
    const dow = date.dayOfWeek;
    return dow !== 6 && dow !== 7 && !this.isHoliday(date);
  }

  /**
   * @param {Temporal.PlainDate} date
   * @returns {boolean}
   */
  isHoliday(date) {
    const d = date.toString();
    return (
      this.holidays.includes(d) &&
      !(this.role === 'teacher' && this.teacherWorkDays.includes(d))
    );
  }

  /**
   * Find the next holiday after the given instant.
   * @param {Temporal.Instant} instant
   * @returns {Temporal.PlainDate}
   */
  nextHoliday(instant) {
    let d = instant.toZonedDateTimeISO(this.timezone).toPlainDate().add({ days: 1 });
    while (!this.isHoliday(d) && Temporal.PlainDate.compare(d, this.lastDay) <= 0) {
      d = d.add({ days: 1 });
    }
    return d;
  }

  /**
   * @param {Temporal.Instant} instant
   * @returns {Temporal.Instant}
   */
  nextSchoolDayStart(instant) {
    const date = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    if (this.isSchoolDay(date)) {
      const start = this.schedule(date).startOfDay(date, this.timezone);
      if (Temporal.Instant.compare(start, instant) > 0) {
        return start;
      }
    }
    const next = this.#nextSchoolDay(date);
    return this.schedule(next).startOfDay(next, this.timezone);
  }

  /**
   * @param {Temporal.Instant} instant
   * @returns {Temporal.Instant}
   */
  previousSchoolDayEnd(instant) {
    const date = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    if (this.isSchoolDay(date)) {
      const end = this.schedule(date).endOfDay(date, this.timezone);
      if (Temporal.Instant.compare(end, instant) < 0) {
        return end;
      }
    }
    const prev = this.#previousSchoolDay(date);
    return this.schedule(prev).endOfDay(prev, this.timezone);
  }

  /**
   * @param {Temporal.Instant} instant
   * @returns {Interval | null}
   */
  currentInterval(instant) {
    const date = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    const sched = this.schedule(date);
    return sched.currentInterval(instant);
  }

  /**
   * Count school days remaining from `instant` through end of year.
   * @param {Temporal.Instant} instant
   * @returns {number}
   */
  schoolDaysLeft(instant) {
    const date = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    const sched = this.schedule(date);
    const endOfDay = this.isSchoolDay(date) ? sched.endOfDay(date, this.timezone) : null;

    let count = 0;
    if (endOfDay && Temporal.Instant.compare(instant, endOfDay) < 0) {
      count = 1; // currently a school day, counts as remaining
    }

    // Always start counting from tomorrow regardless.
    let d = date.add({ days: 1 });
    const endDate = this.lastDay;

    while (Temporal.PlainDate.compare(d, endDate) <= 0) {
      if (this.isSchoolDay(d)) count++;
      d = d.add({ days: 1 });
    }
    return count;
  }

  /**
   * Count calendar days remaining until the day after lastDay.
   * @param {Temporal.Instant} instant
   * @returns {number}
   */
  calendarDaysLeft(instant) {
    const date = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    const endDate = this.lastDay.add({ days: 1 });
    return daysBetween(this.#noonInstant(date), this.#noonInstant(endDate));
  }

  /**
   * Compute total school time (as Duration) from `instant` to end of year.
   * @param {Temporal.Instant} instant
   * @returns {Temporal.Duration}
   */
  schoolTimeLeft(instant) {
    return this.#schoolTimeBetween(instant, this.endOfYear());
  }

  /**
   * Compute school time (as Duration) from start of year to `instant`.
   * @param {Temporal.Instant} instant
   * @returns {Temporal.Duration}
   */
  schoolTimeDone(instant) {
    return this.#schoolTimeBetween(this.startOfYear(), instant);
  }

  /**
   * Total school time in the year.
   * @returns {Temporal.Duration}
   */
  totalSchoolTime() {
    return this.#schoolTimeBetween(this.startOfYear(), this.endOfYear());
  }

  /**
   * Public entry point for cross-instance school-time computation.
   * Clamps to this calendar's bounds before summing.
   * @param {Temporal.Instant} start
   * @param {Temporal.Instant} end
   * @returns {Temporal.Duration}
   */
  schoolTimeBetween(start, end) {
    return this.#schoolTimeBetween(start, end);
  }
}

class Schedule {
  /**
   * @param {Calendar} calendar
   * @param {Array<{name: string, start: Temporal.PlainTime, end: Temporal.PlainTime, tags?: string[], teachers?: boolean}>} periods
   * @param {Temporal.PlainDate} date
   */
  constructor(calendar, periods, date) {
    this.calendar = calendar;
    this.date = date;
    this.rawPeriods = periods.map((x) => new Period(x.name, x.start, x.end, x.tags, x.teachers));

    // Set .next links on actual periods.
    const actual = this.actualPeriods();
    actual.forEach((p, i) => {
      p.next = i < actual.length - 1 ? actual[i + 1] : null;
    });
  }

  #maybeBreak(instant) {
    if (this.notInSchool(instant)) {
      const prev = this.calendar.previousSchoolDayEnd(instant);
      const next = this.calendar.nextSchoolDayStart(instant);
      const days = daysBetween(prev, next);
      if (days >= 3) {
        const name = this.#breakName(days, prev, next);
        return new Interval(`${name}!`, prev, next, false, 'break', []);
      }
    }
    return null;
  }

  #breakName(days, start, end) {
    const tz = this.calendar.timezone;
    if (days > 4) {
      const nextHoliday = this.calendar.nextHoliday(start);
      return this.calendar.breakNames[nextHoliday.toString()] || 'Vacation';
    } else if (includesWeekend(start, end, tz)) {
      return days > 3 ? 'Long weekend' : 'Weekend';
    } else {
      return 'Mid-week vacation?';
    }
  }

  /**
   * Determine if a period should be included given the current date/config.
   * @param {Period} p
   * @returns {boolean}
   */
  hasPeriod(p) {
    if (p.teachers) {
      return this.calendar.role === 'teacher';
    }

    const tags = p.tags || [];
    if (!tags.includes('optional')) {
      // Not optional — always include.
      return true;
    }

    // Optional — include only if one of the other tags appears in includeTags for this day.
    const dow = this.date.dayOfWeek; // 1=Mon…7=Sun
    const allowed = this.calendar.includeTags[dow] || [];
    return tags.some((tag) => tag !== 'optional' && allowed.includes(tag));
  }

  /**
   * @returns {Period[]}
   */
  actualPeriods() {
    const base = this.rawPeriods.filter((p) => this.hasPeriod(p));

    if (base.length === 0) return base;

    // Trim optional periods from start and end.
    while (base.length > 0 && (base[0].tags || []).includes('optional')) base.shift();
    while (base.length > 0 && (base[base.length - 1].tags || []).includes('optional')) base.pop();

    return base;
  }

  firstPeriod() {
    return this.actualPeriods()[0];
  }

  lastPeriod() {
    const ps = this.actualPeriods();
    return ps[ps.length - 1];
  }

  /**
   * @param {Temporal.PlainDate} date
   * @param {string} timezone
   * @returns {Temporal.Instant}
   */
  startOfDay(date, timezone) {
    return this.firstPeriod().startInstant(date, timezone);
  }

  /**
   * @param {Temporal.PlainDate} date
   * @param {string} timezone
   * @returns {Temporal.Instant}
   */
  endOfDay(date, timezone) {
    return this.lastPeriod().endInstant(date, timezone);
  }

  /**
   * @param {Temporal.Instant} instant
   * @returns {boolean}
   */
  notInSchool(instant) {
    const date = this.date;
    const tz = this.calendar.timezone;
    return (
      !this.calendar.isSchoolDay(date) ||
      Temporal.Instant.compare(instant, this.endOfDay(date, tz)) >= 0 ||
      Temporal.Instant.compare(instant, this.startOfDay(date, tz)) <= 0
    );
  }

  /**
   * @param {Temporal.Instant} instant
   * @returns {Interval | null}
   */
  currentInterval(instant) {
    const daysOff = this.#maybeBreak(instant);
    if (daysOff) return daysOff;

    const tz = this.calendar.timezone;
    const date = this.date;
    const first = this.firstPeriod();
    const last = this.lastPeriod();

    if (!first) return null;

    if (first.isAfter(instant, date, tz)) {
      return new Interval(
        'Before school',
        this.calendar.previousSchoolDayEnd(instant),
        first.startInstant(date, tz),
        false,
        'before-school',
        [],
      );
    } else if (last.isBefore(instant, date, tz)) {
      return new Interval(
        'After school',
        last.endInstant(date, tz),
        this.calendar.nextSchoolDayStart(instant),
        false,
        'after-school',
        [],
      );
    } else {
      for (let p = first; p !== null; p = p.next) {
        if (p.contains(instant, date, tz)) {
          return p.toInterval(date, tz);
        } else if (p.next && p.isBefore(instant, date, tz) && p.next.isAfter(instant, date, tz)) {
          return new Interval(
            `Passing to ${p.next.name}`,
            p.endInstant(date, tz),
            p.next.startInstant(date, tz),
            true,
            'passing',
            [],
          );
        }
      }
    }

    return null;
  }
}

class Period {
  /**
   * @param {string} name
   * @param {Temporal.PlainTime} start
   * @param {Temporal.PlainTime} end
   * @param {string[] | undefined} tags
   * @param {boolean | undefined} teachers
   */
  constructor(name, start, end, tags, teachers) {
    this.name = name;
    this.start = start; // Temporal.PlainTime
    this.end = end;     // Temporal.PlainTime
    this.tags = tags || [];
    this.teachers = !!teachers;
    this.next = null;
  }

  /**
   * @param {Temporal.PlainDate} date
   * @param {string} timezone
   * @returns {Temporal.Instant}
   */
  startInstant(date, timezone) {
    return date.toPlainDateTime(this.start).toZonedDateTime(timezone).toInstant();
  }

  /**
   * @param {Temporal.PlainDate} date
   * @param {string} timezone
   * @returns {Temporal.Instant}
   */
  endInstant(date, timezone) {
    return date.toPlainDateTime(this.end).toZonedDateTime(timezone).toInstant();
  }

  /**
   * @param {Temporal.Instant} instant
   * @param {Temporal.PlainDate} date
   * @param {string} timezone
   * @returns {boolean}
   */
  isAfter(instant, date, timezone) {
    return Temporal.Instant.compare(this.startInstant(date, timezone), instant) > 0;
  }

  /**
   * @param {Temporal.Instant} instant
   * @param {Temporal.PlainDate} date
   * @param {string} timezone
   * @returns {boolean}
   */
  isBefore(instant, date, timezone) {
    return Temporal.Instant.compare(this.endInstant(date, timezone), instant) < 0;
  }

  /**
   * @param {Temporal.Instant} instant
   * @param {Temporal.PlainDate} date
   * @param {string} timezone
   * @returns {boolean}
   */
  contains(instant, date, timezone) {
    return (
      Temporal.Instant.compare(this.startInstant(date, timezone), instant) < 0 &&
      Temporal.Instant.compare(instant, this.endInstant(date, timezone)) < 0
    );
  }

  /**
   * @param {Temporal.PlainDate} date
   * @param {string} timezone
   * @returns {Interval}
   */
  toInterval(date, timezone) {
    return new Interval(
      this.name,
      this.startInstant(date, timezone),
      this.endInstant(date, timezone),
      true,
      'period',
      this.tags,
    );
  }
}

class Interval {
  /**
   * @param {string} name
   * @param {Temporal.Instant} start
   * @param {Temporal.Instant} end
   * @param {boolean} duringSchool
   * @param {'period'|'passing'|'before-school'|'after-school'|'break'} type
   * @param {string[]} tags
   */
  constructor(name, start, end, duringSchool, type, tags) {
    this.name = name;
    this.start = start;
    this.end = end;
    this.duringSchool = duringSchool;
    this.type = type;
    this.tags = tags;
  }

  /**
   * @param {Temporal.Instant} [now]
   * @returns {Temporal.Duration}
   */
  left(now = Temporal.Now.instant()) {
    return now.until(this.end);
  }

  /**
   * @param {Temporal.Instant} [now]
   * @returns {Temporal.Duration}
   */
  done(now = Temporal.Now.instant()) {
    return this.start.until(now);
  }
}

export { Calendar, Schedule, Period, Interval, normalizeIncludeTags };

File: lib-py/bells/calendar.py
"""Calendar, Schedule, Period, and Interval classes.

Python counterpart of the JavaScript ``calendar.js`` module. No global state;
all date/time logic lives in :mod:`bells.datetimeutil`.
"""

from __future__ import annotations

from datetime import date, datetime, time, timedelta
from typing import Optional, Union

from .datetimeutil import (
    days_between,
    includes_weekend,
    instant_to_date,
    now_instant,
    parse_plain_date,
    plain_to_instant,
    resolve_schedule_times,
)

WEEKDAY_NAMES = {
    1: "monday",
    2: "tuesday",
    3: "wednesday",
    4: "thursday",
    5: "friday",
    6: "saturday",
    7: "sunday",
}


def normalize_include_tags(
    include_tags: Union[list[str], dict, None]
) -> dict[int, list[str]]:
    """Normalize the ``include_tags`` option.

    Accepts either a flat list (same tags every weekday) or a per-day-of-week
    map keyed by ISO weekday (1=Mon … 5=Fri). Returns a dict keyed by weekday.
    """
    if isinstance(include_tags, list):
        return {1: include_tags, 2: include_tags, 3: include_tags, 4: include_tags, 5: include_tags}
    return include_tags or {}


class Period:
    """A single named period with wall-clock start/end times."""

    def __init__(
        self,
        name: str,
        start: time,
        end: time,
        tags: Optional[list[str]] = None,
        teachers: bool = False,
    ):
        self.name = name
        self.start = start  # datetime.time
        self.end = end  # datetime.time
        self.tags = tags or []
        self.teachers = bool(teachers)
        self.next: Optional["Period"] = None

    def start_instant(self, d: date, timezone: str) -> datetime:
        return plain_to_instant(d, self.start, timezone)

    def end_instant(self, d: date, timezone: str) -> datetime:
        return plain_to_instant(d, self.end, timezone)

    def is_after(self, instant: datetime, d: date, timezone: str) -> bool:
        return self.start_instant(d, timezone) > instant

    def is_before(self, instant: datetime, d: date, timezone: str) -> bool:
        return self.end_instant(d, timezone) < instant

    def contains(self, instant: datetime, d: date, timezone: str) -> bool:
        return self.start_instant(d, timezone) < instant < self.end_instant(d, timezone)

    def to_interval(self, d: date, timezone: str) -> "Interval":
        return Interval(
            self.name,
            self.start_instant(d, timezone),
            self.end_instant(d, timezone),
            True,
            "period",
            self.tags,
        )


class Interval:
    """A span of time of a particular ``type`` (period, passing, break, …)."""

    def __init__(
        self,
        name: str,
        start: datetime,
        end: datetime,
        during_school: bool,
        type: str,
        tags: list[str],
    ):
        self.name = name
        self.start = start
        self.end = end
        self.during_school = during_school
        self.type = type
        self.tags = tags

    def left(self, now: Optional[datetime] = None) -> timedelta:
        if now is None:
            now = now_instant()
        return self.end - now

    def done(self, now: Optional[datetime] = None) -> timedelta:
        if now is None:
            now = now_instant()
        return now - self.start


class Schedule:
    """A resolved set of periods for one specific date."""

    def __init__(self, calendar: "Calendar", periods: list[dict], d: date, name: Optional[str] = None):
        self.calendar = calendar
        self.date = d
        self.name = name
        self.raw_periods = [
            Period(p["name"], p["start"], p["end"], p.get("tags"), p.get("teachers", False))
            for p in periods
        ]

        # Set .next links on actual periods.
        actual = self.actual_periods()
        for i, p in enumerate(actual):
            p.next = actual[i + 1] if i < len(actual) - 1 else None

    def _maybe_break(self, instant: datetime) -> Optional[Interval]:
        if self.not_in_school(instant):
            prev = self.calendar.previous_school_day_end(instant)
            nxt = self.calendar.next_school_day_start(instant)
            days = days_between(prev, nxt)
            if days >= 3:
                name = self._break_name(days, prev, nxt)
                return Interval(f"{name}!", prev, nxt, False, "break", [])
        return None

    def _break_name(self, days: int, start: datetime, end: datetime) -> str:
        tz = self.calendar.timezone
        if days > 4:
            next_holiday = self.calendar.next_holiday(start)
            return self.calendar.break_names.get(next_holiday.isoformat(), "Vacation")
        elif includes_weekend(start, end, tz):
            return "Long weekend" if days > 3 else "Weekend"
        else:
            return "Mid-week vacation?"

    def has_period(self, p: Period) -> bool:
        """Whether a period should be included given the current date/config."""
        if p.teachers:
            return self.calendar.role == "teacher"

        tags = p.tags or []
        if "optional" not in tags:
            # Not optional — always include.
            return True

        # Optional — include only if one of its other tags appears in
        # include_tags for this day.
        dow = self.date.isoweekday()  # 1=Mon … 7=Sun
        allowed = self.calendar.include_tags.get(dow, [])
        return any(tag != "optional" and tag in allowed for tag in tags)

    def actual_periods(self) -> list[Period]:
        base = [p for p in self.raw_periods if self.has_period(p)]

        if not base:
            return base

        # Trim nonschool optional periods from start and end. These are
        # administrative periods (e.g. Food Trucks) that should not define
        # school day boundaries. User-configurable optional periods (zero,
        # seventh, ext) are kept so that enabling them correctly affects the
        # start/end of the school day.
        def is_nonschool(p: Period) -> bool:
            return "nonschool" in (p.tags or [])

        while base and "optional" in (base[0].tags or []) and is_nonschool(base[0]):
            base.pop(0)
        while base and "optional" in (base[-1].tags or []) and is_nonschool(base[-1]):
            base.pop()

        return base

    def first_period(self) -> Optional[Period]:
        ps = self.actual_periods()
        return ps[0] if ps else None

    def last_period(self) -> Optional[Period]:
        ps = self.actual_periods()
        return ps[-1] if ps else None

    def start_of_day(self, d: date, timezone: str) -> datetime:
        return self.first_period().start_instant(d, timezone)

    def end_of_day(self, d: date, timezone: str) -> datetime:
        return self.last_period().end_instant(d, timezone)

    def not_in_school(self, instant: datetime) -> bool:
        d = self.date
        return (
            not self.calendar.is_school_day(d)
            or instant >= self.end_of_day(d, self.calendar.timezone)
            or instant <= self.start_of_day(d, self.calendar.timezone)
        )

    def current_interval(self, instant: datetime) -> Optional[Interval]:
        days_off = self._maybe_break(instant)
        if days_off:
            return days_off

        tz = self.calendar.timezone
        d = self.date
        first = self.first_period()
        last = self.last_period()

        if first is None:
            return None

        if first.is_after(instant, d, tz):
            return Interval(
                "Before school",
                self.calendar.previous_school_day_end(instant),
                first.start_instant(d, tz),
                False,
                "before-school",
                [],
            )
        elif last.is_before(instant, d, tz):
            return Interval(
                "After school",
                last.end_instant(d, tz),
                self.calendar.next_school_day_start(instant),
                False,
                "after-school",
                [],
            )
        else:
            p = first
            while p is not None:
                if p.contains(instant, d, tz):
                    return p.to_interval(d, tz)
                elif (
                    p.next
                    and p.is_before(instant, d, tz)
                    and p.next.is_after(instant, d, tz)
                ):
                    return Interval(
                        f"Passing to {p.next.name}",
                        p.end_instant(d, tz),
                        p.next.start_instant(d, tz),
                        True,
                        "passing",
                        [],
                    )
                p = p.next

        return None


class Calendar:
    """A single academic year's calendar data."""

    def __init__(self, data: dict, options: dict):
        self.data = data
        self.timezone = data["timezone"]
        self.role = options["role"]
        self.include_tags = normalize_include_tags(options["include_tags"])

        if self.role == "teacher" and data.get("firstDayTeachers"):
            self.first_day = parse_plain_date(data["firstDayTeachers"])
        else:
            self.first_day = parse_plain_date(data["firstDay"])
        self.last_day = parse_plain_date(data["lastDay"])
        self.schedules = data["schedules"]
        self.weekday_schedules = data.get("weekdaySchedules") or {}
        self.dates = data.get("dates") or {}
        self.holidays = data.get("holidays") or []
        self.teacher_work_days = data.get("teacherWorkDays") or []
        self.break_names = data.get("breakNames") or {}
        self.non_class_days = data.get("nonClassDays") or {}

    # ── internal helpers ────────────────────────────────────────────────────

    def _noon_instant(self, d: date) -> datetime:
        return plain_to_instant(d, time(hour=12), self.timezone)

    def _next_school_day(self, d: date) -> date:
        d = d + timedelta(days=1)
        while not self.is_school_day(d):
            d = d + timedelta(days=1)
        return d

    def _previous_school_day(self, d: date) -> date:
        d = d - timedelta(days=1)
        while not self.is_school_day(d):
            d = d - timedelta(days=1)
        return d

    def _school_time_between(self, start: datetime, end: datetime) -> timedelta:
        # Clamp start/end to calendar bounds.
        cal_start = self.start_of_year()
        cal_end = self.end_of_year()
        cursor = cal_start if start < cal_start else start
        finish = cal_end if end > cal_end else end

        if cursor >= finish:
            return timedelta(0)

        total = timedelta(0)
        cursor_date = instant_to_date(cursor, self.timezone)

        while cursor_date <= self.last_day:
            # Stop once the start of this day is past the finish time.
            day_midnight = plain_to_instant(cursor_date, time(hour=0), self.timezone)
            if day_midnight >= finish:
                break

            if self.is_school_day(cursor_date):
                sched = self.schedule(cursor_date)
                day_start = sched.start_of_day(cursor_date, self.timezone)
                day_end = sched.end_of_day(cursor_date, self.timezone)

                frm = cursor if cursor > day_start else day_start
                to = finish if finish < day_end else day_end

                if frm < to:
                    total += to - frm

            cursor_date = cursor_date + timedelta(days=1)

        return total

    def _named_schedule(self, name: str) -> list[dict]:
        periods = self.schedules.get(name)
        if periods is None:
            raise ValueError(f'Unknown schedule "{name}"')
        return periods

    # ── public API ──────────────────────────────────────────────────────────

    def is_in_calendar(self, instant: datetime) -> bool:
        return self.start_of_year() <= instant <= self.end_of_year()

    def start_of_year(self) -> datetime:
        sched = self.schedule(self.first_day)
        return sched.first_period().start_instant(self.first_day, self.timezone)

    def end_of_year(self) -> datetime:
        sched = self.schedule(self.last_day)
        return sched.last_period().end_instant(self.last_day, self.timezone)

    def schedule(self, d: date) -> Schedule:
        ds = d.isoformat()
        name = None
        if ds in self.dates and d >= self.first_day:
            entry = self.dates[ds]
            if isinstance(entry, str):
                periods = self._named_schedule(entry)
                name = entry
            else:
                periods = entry
        else:
            weekday_name = WEEKDAY_NAMES[d.isoweekday()]
            name = self.weekday_schedules.get(weekday_name, "NORMAL")
            periods = self._named_schedule(name)
        return Schedule(self, resolve_schedule_times(periods), d, name)

    def is_school_day(self, d: date) -> bool:
        dow = d.isoweekday()
        return dow != 6 and dow != 7 and not self.is_holiday(d)

    def is_holiday(self, d: date) -> bool:
        ds = d.isoformat()
        return ds in self.holidays and not (
            self.role == "teacher" and ds in self.teacher_work_days
        )

    def next_holiday(self, instant: datetime) -> date:
        d = instant_to_date(instant, self.timezone) + timedelta(days=1)
        while not self.is_holiday(d) and d <= self.last_day:
            d = d + timedelta(days=1)
        return d

    def next_school_day_start(self, instant: datetime) -> datetime:
        d = instant_to_date(instant, self.timezone)
        if self.is_school_day(d):
            start = self.schedule(d).start_of_day(d, self.timezone)
            if start > instant:
                return start
        nxt = self._next_school_day(d)
        return self.schedule(nxt).start_of_day(nxt, self.timezone)

    def previous_school_day_end(self, instant: datetime) -> datetime:
        d = instant_to_date(instant, self.timezone)
        if self.is_school_day(d):
            end = self.schedule(d).end_of_day(d, self.timezone)
            if end < instant:
                return end
        prev = self._previous_school_day(d)
        return self.schedule(prev).end_of_day(prev, self.timezone)

    def current_interval(self, instant: datetime) -> Optional[Interval]:
        d = instant_to_date(instant, self.timezone)
        sched = self.schedule(d)
        return sched.current_interval(instant)

    def school_days_left(self, instant: datetime) -> int:
        d = instant_to_date(instant, self.timezone)
        sched = self.schedule(d)
        end_of_day = sched.end_of_day(d, self.timezone) if self.is_school_day(d) else None

        count = 0
        if end_of_day and instant < end_of_day:
            count = 1  # currently a school day, counts as remaining

        # Always start counting from tomorrow regardless.
        cur = d + timedelta(days=1)
        while cur <= self.last_day:
            if self.is_school_day(cur):
                count += 1
            cur = cur + timedelta(days=1)
        return count

    def non_class_label(self, d: date) -> Optional[str]:
        return self.non_class_days.get(d.isoformat())

    def non_class_days_left(self, instant: datetime) -> list[dict]:
        today = instant_to_date(instant, self.timezone)
        today_sched = self.schedule(today) if self.is_school_day(today) else None
        today_end = today_sched.end_of_day(today, self.timezone) if today_sched else None
        includes_today = bool(today_end and instant < today_end)

        result = []
        for date_str, label in self.non_class_days.items():
            d = parse_plain_date(date_str)
            if d < today:
                continue
            if d == today and not includes_today:
                continue
            if d > self.last_day:
                continue
            result.append({"date": d, "label": label})
        result.sort(key=lambda x: x["date"])
        return result

    def school_days_between(self, start: date, end: date) -> int:
        # Clamp to calendar bounds.
        frm = self.first_day if start < self.first_day else start
        to = self.last_day if end > self.last_day else end

        count = 0
        d = frm
        while d <= to:
            if self.is_school_day(d):
                count += 1
            d = d + timedelta(days=1)
        return count

    def calendar_days_left(self, instant: datetime) -> int:
        d = instant_to_date(instant, self.timezone)
        end_date = self.last_day + timedelta(days=1)
        return days_between(self._noon_instant(d), self._noon_instant(end_date))

    def school_time_left(self, instant: datetime) -> timedelta:
        return self._school_time_between(instant, self.end_of_year())

    def school_time_done(self, instant: datetime) -> timedelta:
        return self._school_time_between(self.start_of_year(), instant)

    def total_school_time(self) -> timedelta:
        return self._school_time_between(self.start_of_year(), self.end_of_year())

    def school_time_between(self, start: datetime, end: datetime) -> timedelta:
        return self._school_time_between(start, end)
